### ISSUE TITLE
declare $var=$val triggers "Don't use $ on the left side of assignments."

### ISSUE THREAD

**Issue by koalaman** [koalaman/shellcheck#843]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC1066
- My shellcheck version (`shellcheck --version` or "online"): online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [ ] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
declare $var=$val
```

#### Here's what shellcheck currently says:

```
declare $var=$val
         ^-- SC1066: Don't use $ on the left side of assignments.
```
#### Here's what I wanted or expected to see:

Nothing or quote warning.

**Comment by kurahaupo** [comment]
On the contrary, anything that involves indirect variable naming should be
warned unless there's an explicit "don't warn" override in the code.
Symbolic indirection is fraught with dangers, special cases, etc.

What about if `$var` is not actually a valid name? What if it's readable but
not writable? What if it refers to a special var?

For example, what happens if `$var` happens to be IFS or PATH or TZ or
LD_LIBRARY_PATH or EUID or ...?

Yes you could put in guards, but did you check for PATH[anything] ?
Did you check for anything[IFS++] ?

Really it's worth avoiding just on principle.

There are almost always better ways to achieve the same effect -- and 99
times out of 100, that's by using an associative array.

What are you actually trying to do with that code?

**Comment by koalaman** [comment]
This warning is purely a bug caused by reusing the lenient assignment parser for `define`, `export` and friends. It has other unintentional effects like not warning about quoting for $var/$val, and this thing:

```
declare n$1=foo
        ^-- SC1067: For indirection, use (associative) arrays or 'read "var$n" <<< "value"'
```

Indirect assignment might be bad, but a warning about an unrelated issue doesn't help that.

This has been fixed in 2688a81
